File: dataprovider-retrofit/src/main/java/ch/srg/dataProvider/integrationlayer/utils/StreamComparator.java
package ch.srg.dataProvider.integrationlayer.utils;

import androidx.annotation.Nullable;

import java.util.Comparator;
import java.util.List;

import ch.srg.dataProvider.integrationlayer.data.Quality;
import ch.srg.dataProvider.integrationlayer.data.Resource;
import ch.srg.dataProvider.integrationlayer.data.StreamingMethod;

/**
 * Copyright (c) SRG SSR. All rights reserved.
 * <p>
 * License information is available from the LICENSE file.
 */
public class StreamComparator implements Comparator<Resource> {
    private static final int HANDICAP_STREAMING = 1000;
    private static final int HANDICAP_DVR = 10;
    private static final int HANDICAP_QUALITY = 1;

    public static final int SCORE_NOT_SUPPORTED = Integer.MAX_VALUE;

    private List<Quality> orderedQualities;
    private List<StreamingMethod> orderedStreaming;
    @Nullable
    private List<Resource.Drm.Type> supportedDrms;
    private boolean dvrSupported;

    public StreamComparator(List<Quality> orderedQualities, List<StreamingMethod> orderedStreaming, @Nullable List<Resource.Drm.Type> supportedDrms, boolean dvrSupported) {
        this.orderedQualities = orderedQualities;
        this.orderedStreaming = orderedStreaming;
        this.supportedDrms = supportedDrms;
        this.dvrSupported = dvrSupported;
    }

    @Override
    public int compare(Resource lhs, Resource rhs) {
        int l = score(lhs);
        int r = score(rhs);
        //noinspection UseCompareMethod
        return l == r ? 0 : ((l > r) ? 1 : -1);
    }

    public int score(Resource r) {
        if (r.getStreamingMethod() == null
                || (!dvrSupported && r.getDvr())
                || (!isSupportedDrm(r.getDrmList()))) {
            return SCORE_NOT_SUPPORTED;
        }
        int indexOfStreaming = orderedStreaming.indexOf(r.getStreamingMethod()); // will retrun -1 if not in the list
        int indexOfQuality = orderedQualities.indexOf(r.getQuality()); // will retrun -1 if not in the list

        return (indexOfStreaming < 0 ? orderedStreaming.size() : indexOfStreaming) * HANDICAP_STREAMING
                + (indexOfQuality < 0 ? orderedQualities.size() : indexOfQuality) * HANDICAP_QUALITY
                + (r.getDvr() ? 0 : HANDICAP_DVR);
    }

    private boolean isSupportedDrm(List<Resource.Drm> drmList) {
        if (drmList == null) {
            return true;
        }
        if (supportedDrms != null) {
            for (Resource.Drm drm : drmList) {
                if (supportedDrms.contains(drm.getType())) {
                    return true;
                }
            }
        }
        return false;
    }
}

File: dataprovider-retrofit/src/main/java/ch/srg/dataProvider/integrationlayer/utils/StreamComparator.kt
package ch.srg.dataProvider.integrationlayer.utils

import ch.srg.dataProvider.integrationlayer.data.remote.Quality
import ch.srg.dataProvider.integrationlayer.data.remote.Resource
import ch.srg.dataProvider.integrationlayer.data.remote.StreamingMethod

/**
 * Copyright (c) SRG SSR. All rights reserved.
 * <p>
 * License information is available from the LICENSE file.
 */
class StreamComparator(
    private val orderedQualities: List<Quality>,
    private val orderedStreaming: List<StreamingMethod>,
    private val supportedDrms: List<Resource.Drm.Type>?,
    private val dvrSupported: Boolean,
) : Comparator<Resource> {
    override fun compare(lhs: Resource, rhs: Resource): Int {
        val l = score(lhs)
        val r = score(rhs)

        return l.compareTo(r)
    }

    fun score(resource: Resource): Int {
        if ((!dvrSupported && resource.dvr) || (!isSupportedDrm(resource.drmList))) {
            return SCORE_NOT_SUPPORTED
        }

        val indexOfStreaming = orderedStreaming.indexOf(resource.streamingMethod) // Will return -1 if not in the list
        val indexOfQuality = orderedQualities.indexOf(resource.quality) // Will return -1 if not in the list
        val scoreStreaming = (if (indexOfStreaming < 0) orderedStreaming.size else indexOfStreaming) * HANDICAP_STREAMING
        val scoreQuality = (if (indexOfQuality < 0) orderedQualities.size else indexOfQuality) * HANDICAP_QUALITY
        val scoreDvr = if (resource.dvr) 0 else HANDICAP_DVR

        return scoreStreaming + scoreQuality + scoreDvr
    }

    private fun isSupportedDrm(drmList: List<Resource.Drm>?): Boolean {
        if (drmList == null) {
            return true
        }

        if (supportedDrms.isNullOrEmpty()) {
            return false
        }

        return drmList.any { it.type in supportedDrms }
    }

    companion object {
        const val SCORE_NOT_SUPPORTED = Int.MAX_VALUE

        private const val HANDICAP_STREAMING = 1000
        private const val HANDICAP_DVR = 10
        private const val HANDICAP_QUALITY = 1
    }
}
